Userland/Applications/FileManager/PropertiesWindowHashesTab.cpp

```diff
@@ -0,0 +1,184 @@
+    VERIFY(!m_icon_hash_correct.is_null());
+    m_icon_hash_incorrect = Gfx::Bitmap::try_load_from_file("/res/icons/16x16/hash-incorrect.png");
+    VERIFY(!m_icon_hash_incorrect.is_null());
+    m_icon_hash_load = Gfx::Bitmap::try_load_from_file("/res/icons/16x16/hash-ask.png");
+    VERIFY(!m_icon_hash_load.is_null());
+    m_icon_hash = Gfx::Bitmap::try_load_from_file("/res/icons/16x16/hash.png");
+    VERIFY(!m_icon_hash.is_null());
+    m_icon_copy = Gfx::Bitmap::try_load_from_file("/res/icons/16x16/edit-copy.png");
+    VERIFY(!m_icon_copy.is_null());
+
+    m_md5_widget = add<HashWidget>(this->make_weak_ptr<HashTab>(), String("MD5:"), Crypto::Hash::HashKind::MD5);
+    m_sha1_widget = add<HashWidget>(this->make_weak_ptr<HashTab>(), String("SHA1:"), Crypto::Hash::HashKind::SHA1);
+    m_sha256_widget = add<HashWidget>(this->make_weak_ptr<HashTab>(), String("SHA256:"), Crypto::Hash::HashKind::SHA256);
+    m_sha512_widget = add<HashWidget>(this->make_weak_ptr<HashTab>(), String("SHA512:"), Crypto::Hash::HashKind::SHA512);
+}
+
+HashTab::~HashTab()
+{
+}
+
+void HashTab::show_event(GUI::ShowEvent&)
+{
+    if (m_calculate_on_show) {
+        m_md5_widget->background_calculate_set_and_update(this->make_weak_ptr<HashTab>());
+        m_sha1_widget->background_calculate_set_and_update(this->make_weak_ptr<HashTab>());
+        m_sha256_widget->background_calculate_set_and_update(this->make_weak_ptr<HashTab>());
+        m_sha512_widget->background_calculate_set_and_update(this->make_weak_ptr<HashTab>());
+    }
+    GUI::Clipboard::the().on_change = [this](const String& mime_type) -> void {
+        if (mime_type.starts_with("text/")) {
+            this->update_clipboard();
+            this->m_md5_widget->update_status(this->make_weak_ptr<HashTab>());
+            this->m_sha1_widget->update_status(this->make_weak_ptr<HashTab>());
+            this->m_sha256_widget->update_status(this->make_weak_ptr<HashTab>());
+            this->m_sha512_widget->update_status(this->make_weak_ptr<HashTab>());
+        }
+    };
+}
+
+Optional<String> HashTab::hash_text_for_file(const Crypto::Hash::HashKind hash_kind) const
+{
+    String path = m_path;
+    auto file_to_hash = Core::File::construct(path);
+    Crypto::Hash::Manager hash;
+    hash.initialize(hash_kind);
+    String hash_out_string;
+    if (file_to_hash->open(Core::OpenMode::ReadOnly)) {
+        while (!file_to_hash->eof() && !file_to_hash->has_error())
+            hash.update(file_to_hash->read(PAGE_SIZE));
+        auto digest = hash.digest();
+        auto digest_data = digest.immutable_data();
+        StringBuilder hash_string_builder;
+        for (size_t i = 0; i < hash.digest_size(); i++)
+            hash_string_builder.appendff("{:02x}", digest_data[i]);
+        hash_out_string = hash_string_builder.build();
+        file_to_hash->close();
+    } else {
+        return Optional<String>();
+    }
```

Comment:
1. Please use fail-early code instead of putting the "else return" at the end
2. I _think_ you can use `return {};` instead of `return Optional<String>();`, which is how it's done in most of the system. In case you can't for some technical reason, discard this.

```suggestion
    if (!file_to_hash->open(Core::OpenMode::ReadOnly)) {
        return {};
    }
    
    while (!file_to_hash->eof() && !file_to_hash->has_error())
        hash.update(file_to_hash->read(PAGE_SIZE));
    auto digest = hash.digest();
    auto digest_data = digest.immutable_data();
    StringBuilder hash_string_builder;
    for (size_t i = 0; i < hash.digest_size(); i++)
        hash_string_builder.appendff("{:02x}", digest_data[i]);
    hash_out_string = hash_string_builder.build();
    file_to_hash->close();
```